### ISSUE TITLE
`disallowed_macros` false negatives

### ISSUE THREAD

**Issue by ojeda** [rust-lang/rust-clippy#11431]
### Summary

`disallowed_macros` has several false negatives. See reproducer.

### Lint Name

`disallowed_macros`

### Reproducer

With `clippy.toml`:

```toml
disallowed-macros = [ "x::m" ]
```

And a command line like:

```sh
clippy-driver --edition=2021 -Dclippy::disallowed_macros x.rs
```

This works (similar to an existing test in Clippy):

```rust
macro_rules! m (
    ($t:tt) => (1)
);

fn main() {
    let _ = m!(42);
}
```

However, this does not:

```rust
macro_rules! m (
    ($t:tt) => ($t)
);

fn main() {
    let _ = m!(42);
}
```

Nor:

```rust
macro_rules! m (
    ($t:tt) => ($t + $t)
);

fn main() {
    let _ = m!(42);
}
```

In addition, empty expansions also do not work -- this may be due to the lint not having a pre-expansion part, but still, it would be nice to handle:

```rust
macro_rules! m (
    ($t:tt) => ()
);

fn main() {
    m!(42);
}
```

### Version

```text
rustc 1.72.0 (5680fa18f 2023-08-23)
rustc 1.74.0-nightly (84a9f4c6e 2023-08-29)
```

**Comment by Alexendoo** [comment]
Opened #11439 for `($t:tt) => ($t + $t)`

`($t:tt) => ($t)` doesn't get detected because it doesn't produce any of its own tokens, from the HIR's perspective `m!(42)` is the same as `42`. A pre expansion pass would be able to see an invocation of it as long it's not part of another macro expansion, but it's a late lint because we need to resolve the macro paths to definitions

`($t:tt) => ()` does produce an empty item in the AST, but those are not preserved in the HIR so we don't pick it up

**Comment by ojeda** [comment]
As an update, [we started using `allow-invalid`](https://github.com/Rust-for-Linux/linux/commit/c016722fd57551f8a6fcf472c9d2bcf2130ea0ec) because (as expected) Clippy has not been finding our custom `dbg!` macro.

**Comment by Alexendoo** [comment]
That is likely https://github.com/rust-lang/rust-clippy/pull/14733, it's scheduled for a backport

Are you still hitting other FNs with it?

**Comment by ojeda** [comment]
> That is likely https://github.com/rust-lang/rust-clippy/pull/14733, it's scheduled for a backport.

Sounds good.

> Are you still hitting other FNs with it?

Hmm... What do you mean?

**Comment by blyxyas** [comment]
Looking deeper into this, I'm not sure if it's actionable. The compiler doesn't emit any distinct information from non-token-emitting macros. It just makes a "no-op". The way that the lint (or any lint for that matter) detects that it's in a macro is via checking for tokens with a [`SyntaxCtxt`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_span/hygiene/struct.SyntaxContext.html) higher than 0.

Being that pre-expansion lints are deprecated any may (will?) be removed in the future, I think the only way forward with this is to fix it in a project-by-project basis.

---

While trying to find the offending occurence, I suppose it was removed or fixed, as the only similar instance of a macro in the kernel I found [was here](https://rust.docs.kernel.org/src/kernel/print.rs.html#177-183) in `print_macro`.

For future occurrences of this false negative, try using `($t:tt) => ({$t})` if it makes sense in the scenario (Note the extra scope)

**Comment by ojeda** [comment]
> Looking deeper into this, I'm not sure if it's actionable. The compiler doesn't emit any distinct information from non-token-emitting macros.

I guess you mean for the empty expansions case (i.e. the last one above: `($t:tt) => ()`), right? Or do you mean the `($t:tt) => ($t)` one too?

> While trying to find the offending occurence, I suppose it was removed or fixed, as the only similar instance of a macro in the kernel I found [was here](https://rust.docs.kernel.org/src/kernel/print.rs.html#177-183) in print_macro.

If I recall correctly, I filled this as a result of trying to debug our `kernel::dbg!` issue with `disallowed-macros`, i.e. I was trying to minimize that one to report it and found more cases, so I reported the false negatives I found here.

The `kernel::dbg!` is really the only one we have so far used by `disallowed-macros`:

```toml
disallowed-macros = [
    # The `clippy::dbg_macro` lint only works with `std::dbg!`, thus we simulate
    # it here, see: https://github.com/rust-lang/rust-clippy/issues/11303.
    { path = "kernel::dbg", reason = "the `dbg!` macro is intended as a debugging tool", allow-invalid = true },
]
```

which in turn is because we cannot use the `dbg_macro` lint since it is a custom one, as the comment mentions.

I hope that clarifies a bit -- thanks for taking a look at this!
